garmin_fit.cc

```diff
@@ -769,6 +880,478 @@ fit_read()
+  { kFieldTimestamp,      0x04, kTypeUint32 },
+  { kFieldEvent,          0x01, kTypeEnum   },
+  { kFieldEventType,      0x01, kTypeEnum   },
+  { kFieldEventGroup,     0x01, kTypeUint8  },
+};
+const static std::vector<fit_field_t> fit_msg_fields_course_point = {
+  { kFieldCPTimeStamp,    0x04, kTypeUint32 },
+  { kFieldCPPositionLat,  0x04, kTypeSint32 },
+  { kFieldCPPositionLong, 0x04, kTypeSint32 },
+  { kFieldCPDistance,     0x04, kTypeUint32 },
+  { kFieldCPName,         0x10, kTypeString },
+  { kFieldCPType,         0x01, kTypeEnum   },
+};
+const static std::vector<fit_field_t> fit_msg_fields_record = {
+  { kFieldTimestamp,      0x04, kTypeUint32 },
+  { kFieldLatitude,       0x04, kTypeSint32 },
+  { kFieldLongitude,      0x04, kTypeSint32 },
+  { kFieldDistance,       0x04, kTypeUint32 },
+  { kFieldAltitude,       0x02, kTypeUint16 },
+  { kFieldSpeed,          0x02, kTypeUint16 },
+};
+
+
+static void
+fit_write_message_def(uint8_t local_id, uint16_t global_id, const std::vector<fit_field_t> &fields) {
+  gbfputc(0x40 | local_id, fout); // Local ID
+  gbfputc(0, fout); // Reserved
+  gbfputc(0, fout); // Little endian
+  gbfputuint16(global_id, fout); // Global ID
+  gbfputc(fields.size(), fout); // Number of fields
+  for (auto &&field : fields) {
+    gbfputc(field.id, fout); // Field definition number
+    gbfputc(field.size, fout); // Field size in bytes
+    gbfputc(field.type, fout); // Field type 
+  }
+}
+
+
+static uint16_t
+fit_crc16(uint8_t data, uint16_t crc) {
+  static const uint16_t crc_table[] = {
+    0x0000, 0xcc01, 0xd801, 0x1400, 0xf001, 0x3c00, 0x2800, 0xe401,
+    0xa001, 0x6c00, 0x7800, 0xb401, 0x5000, 0x9c01, 0x8801, 0x4400
+  };
+  
+  crc = (crc >> 4) ^ crc_table[crc & 0xf] ^ crc_table[data & 0xf];
+  crc = (crc >> 4) ^ crc_table[crc & 0xf] ^ crc_table[(data >> 4) & 0xf];
+  return crc;
+}
+
+
+static void
+fit_write_timestamp(const gpsbabel::DateTime &t) {
+  uint32_t t_fit;
+  if (t.isValid() && t.toTime_t() >= kFitUnixEpochOffset) {
+    t_fit = t.toTime_t() - kFitUnixEpochOffset;
+  } else {
+    t_fit = 0xffffffff;
+  }
+  gbfputuint32(t_fit, fout);
+}
+
+
+static void
+fit_write_fixed_string(const QString &s, unsigned int len) {
+  QString trimmed(s);
+  QByteArray u8buf;
+  
+  // Truncate if too long, making sure not to chop in the middle of a UTF-8
+  // character (i.e. we chop the unicode string and then check whether its
+  // UTF-8 representation fits)
+  while (true) {
+    u8buf = trimmed.toUtf8();
+    if (static_cast<unsigned int>(u8buf.size()) < len) {
+      break;
+    }
+    trimmed.chop(1);
+  }
+  // If the string was too short initially or we had to chop multibyte
+  // characters, the UTF-8 representation might be too short now, so pad
+  // it.
+  u8buf.append(len - u8buf.size(), '\0');
+  gbfwrite(u8buf.data(), len, 1, fout);
+}
+
+
+static void
+fit_write_position(double pos) {
+  if (pos >= -180 && pos < 180) {
+    gbfputint32((pos / 180.0) * 0x7fffffff, fout);
```

Comment:
Please use GPS_Math_Deg_To_Semi instead.

Reply:
Thanks for your feedback, I changed this as requested in the updated pull request.